File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    if(head === null) return null;
    let length = 0;
    let temp = head;

    while(temp !== null){
        temp = temp.next;
        length++;
    }

    if(length === n) return head.next;
    let res = length - n;

    temp = head;

    while(temp !== null){
        res--;
        if(!res) break;
        temp = temp.next;
    }

    temp.next = temp.next.next;
    return head;
};

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if(head.next == null) return null;
        int size = 0;
        ListNode temp = head;
        while(temp != null){
            size++;
            temp = temp.next;
        }

        if(size == n) return head.next;

        int idx = size - n; // 3
        temp = head;
        int cnt = 0;

        while(temp != null){
            cnt++;
            if(cnt == idx){
                temp.next = temp.next.next;
                break;
            }
            temp = temp.next;
        }

        return head;
    }
}